### PR TITLE
Add min and max brightness parameters for Light dim_relative Action

### PR DESCRIPTION
**Pull request by PaoloTK**
# What does this implement/fix?

I have implemented parameters to set the minimum and maximum brightness a dim_relative action can dim a light to.
The most common use case for this feature is to set a min_brightness of 1% (or higher) to disable the light getting turned off as it's dimmed with a physical button, which is common on many traditional and smart dimmers. Rather than limiting the user to this I figured adding min and max brightness parameters would be more versatile. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3984

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
binary_sensor:
  - platform: gpio
    name: "Dim Down"
    id: dim_down
    pin: GPIO34
    icon: "mdi:light-switch"
    on_multi_click: 
    - timing:
        - ON for at most 0.5s
        - OFF for at least 0.5s
      then:
        - light.toggle: dimmable_light
    - timing:
        - ON for at least 0.51s
      then:
        - if:
            condition:
              light.is_on: dimmable_light
            then:
              - while:
                  condition:
                    binary_sensor.is_on: dim_down
                  then:
                    - light.dim_relative:
                        id: dimmable_light
                        relative_brightness: -5%
                        transition_length: 0.1s
                        min_brightness: 1%
                    - delay: 0.1s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by jesserockz — APPROVED** [review]

**Review by jesserockz — CHANGES_REQUESTED** [review]
Please add a config option for what do do in the case the light is already outside the min/max range:

```py
LIMIT_MODES = {
    "CLAMP": LimitMode.CLAMP,
    "DO_NOTHING": LimitMode.DO_NOTHING,
}

...
cv.Optional(CONF_LIMIT_MODE, default="CLAMP"): cv.enum(LIMIT_MODES, upper=True, space="_")
```

**Comment by PaoloTK** [comment]
hi, any news on this?

File: esphome/components/light/automation.h [hunk 1]
```diff
@@ -77,16 +79,29 @@ template<typename... Ts> class DimRelativeAction : public Action<Ts...> {
     float rel = this->relative_brightness_.value(x...);
     float cur;
     this->parent_->remote_values.as_brightness(&cur);
-    float new_brightness = clamp(cur + rel, 0.0f, 1.0f);
+    float new_brightness = clamp(cur + rel, min_brightness_, max_brightness_);
+    if ((limit_mode_ == LimitMode::DO_NOTHING) && ((cur < min_brightness_) || (cur > max_brightness_))) {
+      return;
+    }
```
Comment (jesserockz) [inline esphome/components/light/automation.h]: No point doing the clamping if the value was out of range to begin
```suggestion
    if ((limit_mode_ == LimitMode::DO_NOTHING) && ((cur < min_brightness_) || (cur > max_brightness_))) {
      return;
    }
    float new_brightness = clamp(cur + rel, min_brightness_, max_brightness_);
```

File: esphome/components/light/automation.h [hunk 2]
```diff
@@ -7,6 +7,8 @@
 namespace esphome {
 namespace light {
 
+enum LimitMode { CLAMP, DO_NOTHING };
```
Comment (jesserockz) [inline esphome/components/light/automation.h]:
```suggestion
enum class LimitMode { CLAMP, DO_NOTHING };
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Please add the new config options to a test yaml file

**Comment by PaoloTK** [comment]
Thank you!
I have implemented the requested changes. I have also added max_brightness parameter to tests.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**